source/common/upstream/ring_hash_lb.cc

```diff
@@ -37,41 +34,47 @@ void RingHashLoadBalancer::initialize() {
 
 HostConstSharedPtr
 RingHashLoadBalancer::LoadBalancerImpl::chooseHost(LoadBalancerContext* context) {
-  if (global_panic_) {
+  // Make sure we correctly return nullptr for any early chooseHost() calls.
+  if (per_priority_state_.size() == 0) {
+    return nullptr;
+  }
+  // If there is no hash in the context, just choose a random value (this effectively becomes
+  // the random LB but it won't crash if someone configures it this way).
+  // computeHashKey() may be computed on demand, so get it only once.
+  Optional<uint64_t> hash;
+  if (context) {
+    hash = context->computeHashKey();
+  }
+  const uint64_t h = hash.valid() ? hash.value() : random_.random();
+
+  uint32_t priority = LoadBalancerBase::choosePriority(h, *per_priority_load_);
+  if (per_priority_state_[priority]->global_panic_) {
     stats_.lb_healthy_panic_.inc();
   }
-  return ring_->chooseHost(context, random_);
+  return per_priority_state_[priority]->current_ring_->chooseHost(h);
 }
 
 LoadBalancerPtr RingHashLoadBalancer::LoadBalancerFactoryImpl::create() {
   // We must protect current_ring_ via a RW lock since it is accessed and written to by multiple
-  // threads. All complex processing happens outside of locking however.
-  RingConstSharedPtr ring_to_use;
-  bool global_panic_to_use;
-  {
-    std::shared_lock<std::shared_timed_mutex> lock(mutex_);
-    ring_to_use = current_ring_;
-    global_panic_to_use = global_panic_;
+  // threads. All complex processing has already been precalculated however.
+  auto lb = std::make_unique<LoadBalancerImpl>(stats_, random_);
+
+  std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+  for (size_t i = 0; i < per_priority_state_.size(); ++i) {
+    lb->per_priority_state_.push_back(PerPriorityStatePtr{new PerPriorityState});
```

Comment:
Can you build an allocate the vector outside of the lock? Then you can just acquire the lock and copy the data values in.